src/components/Email/useVerifyEmailOnPage.ts

```diff
@@ -0,0 +1,97 @@
+import { useRouter } from 'next/router';
+import { useEffect } from 'react';
+import { graphql, useFragment } from 'react-relay';
+import { SelectorStoreUpdater } from 'relay-runtime';
+
+import { useToastActions } from '~/contexts/toast/ToastContext';
+import { useVerifyEmailOnPageMutation } from '~/generated/useVerifyEmailOnPageMutation.graphql';
+import { useVerifyEmailOnPageQueryFragment$key } from '~/generated/useVerifyEmailOnPageQueryFragment.graphql';
+import { usePromisifiedMutation } from '~/hooks/usePromisifiedMutation';
+import isFeatureEnabled, { FeatureFlag } from '~/utils/graphql/isFeatureEnabled';
+
+export default function useVerifyEmailOnPage(queryRef: useVerifyEmailOnPageQueryFragment$key) {
+  const query = useFragment(
+    graphql`
+      fragment useVerifyEmailOnPageQueryFragment on Query {
+        viewer {
+          ... on Viewer {
+            email {
+              verificationStatus
+            }
+          }
+        }
+        ...isFeatureEnabledFragment
+      }
+    `,
+    queryRef
+  );
+
+  const verificationStatus = query.viewer?.email?.verificationStatus;
+  const isEmailFeatureEnabled = isFeatureEnabled(FeatureFlag.EMAIL, query);
+
+  const [verifyEmailMutate] = usePromisifiedMutation<useVerifyEmailOnPageMutation>(graphql`
+    mutation useVerifyEmailOnPageMutation($token: String!) @raw_response_type {
+      verifyEmail(token: $token) {
+        ... on VerifyEmailPayload {
+          __typename
+          email
+        }
+        ... on ErrInvalidInput {
+          __typename
+        }
+      }
+    }
+  `);
+
+  const router = useRouter();
+  const { verifyEmail } = router.query;
+  const { pushToast } = useToastActions();
+
+  const updater: SelectorStoreUpdater<useVerifyEmailOnPageMutation['response']> = (
+    store,
+    response
+  ) => {
+    if (response.verifyEmail?.__typename === 'VerifyEmailPayload' && response.verifyEmail?.email) {
+      store
+        .get('client:root:viewer')
+        ?.getLinkedRecord('email')
+        ?.setValue('Verified', 'verificationStatus');
+    }
+  };
```

Comment:
I don't think we need a custom updater here? We should be able to use `optimisticResponse` instead since we're not doing anything too crazy.

`optimisticResponse` should be strongly typed, you can use it like this
```ts
someMutation({
  variables: {},
  optimisticResponse: {
     verifyEmail: {
       // your data here
     }
  }
})
```

As long as you put `@raw_response_type` at the end of your mutation (which you did), the types for the `optimisticResponse` field will be generated for you.

Reply:
@tbezman  Make sense! Since its just a simple updated. 

In this case, we want to update the `verificationStatus` but it doesn't exist inside the `VerifyEmailPayload`

```
type VerifyEmailPayload {
    email: String
}
```

What's your recommend on this? 

1. Should we ask backend to return the verification status, so we can used the `optimisticResponse`
2. Is there any way to update the value of `verificationStatus` from `pending` to `verified` without using `updater`?

The goal is to updated the verification status in this modal (video below):

https://user-images.githubusercontent.com/4480258/201610416-fa50c190-1aa5-4ce4-a109-3e988ad05858.mp4